game/lib.rs

```diff
@@ -10,229 +10,282 @@ mod game {
-    //   * During each round each player gets additional `GAS_CREDITS` added
-    //     to their allowance. This is on top of what they haven't spent last
-    //     time.
-    //   * The value can also be negative, there are penalties for e.g.
-    //     over-painting pixels.
-    type GasAllowance = i32;
-
-    // The amount of gas allowance each player gets credited with each round.
-    const GAS_ALLOWANCE_PER_ROUND: i32 = 750_000_000;
+    /// How much score should be addded per field that is occupied by a user.
+    const SCORE_PER_FIELD: u64 = 1_000_000_000;
 
     #[ink(storage)]
-    pub struct Game {
-        /// Privileged account to control the game.
+    pub struct SquinkSplash {
+        /// Owner of the contract. Can initiate a new game.
         admin: AccountId,
-        /// Mapping of all players.
-        players: Mapping<AccountId, Player>,
-        /// Vector with all players, since we can't iterate over the `players` `Mapping`.
-        all_players: Vec<AccountId>,
-        /// Individual pixels with information about who painted them.
-        pixels: Mapping<(X, Y), Pixel>,
-        /// Playground width.
-        width: Width,
-        /// Playground height.
-        height: Height,
-        /// Keeps track of each players score.
-        _scoreboard: Mapping<PlayerId, Score>,
+        /// In which game phase is this contract.
+        state: State,
+        /// List of fields with their owner (if any).
+        board: Mapping<u32, AccountId>,
+        /// Width and height of the board.
+        dimensions: (u32, u32),
+        /// List of all players.
+        players: Lazy<Vec<Player>>,
+        /// The amount of balance that needs to be payed to join the game.
+        buy_in: Balance,
     }
 
-    #[derive(scale::Decode, scale::Encode)]
+    #[derive(scale::Decode, scale::Encode, Clone)]
     #[cfg_attr(
         feature = "std",
-        derive(
-            Debug,
-            PartialEq,
-            Eq,
-            scale_info::TypeInfo,
-            ink::storage::traits::StorageLayout
-        )
+        derive(Debug, scale_info::TypeInfo, ink::storage::traits::StorageLayout)
     )]
-    struct Player {
-        id: PlayerId,
-        gas_balance: GasAllowance,
+    pub enum State {
+        Forming,
+        Running { end_block: u32 },
+        Finished { winner: AccountId },
     }
 
-    #[derive(Default, scale::Encode, scale::Decode, Eq, PartialEq)]
+    #[derive(scale::Decode, scale::Encode)]
     #[cfg_attr(
         feature = "std",
-        derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+        derive(Debug, scale_info::TypeInfo, ink::storage::traits::StorageLayout)
     )]
-    struct Pixel {
-        owner: Option<PlayerId>,
-    }
-
-    #[ink(event)]
-    pub struct Scoreboard {
-        scoreboard: Vec<(PlayerId, Score)>,
-    }
-
-    #[ink(event)]
-    #[derive(Debug)]
-    pub struct PaintedPixel {
-        player: AccountId,
-        pixel: (X, Y),
+    pub struct Player {
```

Comment:
I think we should user names for the player contract too. It will be easier to identify the players (and yourself). We can do it in another PR. Just mentioning it

Reply:
Good idea. Added.